Complete scored blocks

jcdyer opened the pull request:
This contains the completion logic for scorable xblocks.  When a block is scored, it should be marked complete (completion = 1.0).  If an instructor resets the state of the xblock, it should also reset the completion of the xblock to 0.0.

**JIRA tickets**: N/A

**Discussions**: Architecture discussed extensively on the wiki and in meetings, then the revised proposal was presented to the Arch Council on Oct. 21 and given thumbs up.

**Dependencies**: None

**Screenshots**: N/A

**Sandbox URL**: TBD

**Merge deadline**: None



**Testing instructions**:

You will be going back and forth between the django shell and interacting with the devstack through the browser, so log into the devstack twice, and run `paver devstack lms` in one of the sessions. Then: 

1. Log in as an admin user.
2. In the admin, at http://localhost:18000/admin/grades/persistentgradesenabledflag/add/ select both "enable" and "enable for all courses."
3. Log in as a course staff user.
2. Find a scorable block that the user hasn't answered yet.  Get the block key for that user in the staff debug section.
3. In the django shell (`./manage.py lms --settings=devstack shell` in the other window), ensure that there is no completion object for that block, or if it does exist, ensure the completion value is set to zero.
4. Answer the problem.
5. Refetch the completion object for the block. If you already have a copy, I think you should be able to do `completion_object.refresh_from_db()`.  Otherwise `completion_object = BlockCompletion.objects.get(pk=completion_object.pk)` will work. Ensure the completion value is set to 1.0.
6. In the instructor control panel, reset student state for the given problem and the logged in user.
7. Refetch the completion object again, Ensure the completion value is set to 0.0.

**Reviewers**:

- [x] OpenCraft reviewer: @bradenmacdonald 
- [x] edX reviewer: @robrap

**Author notes and concerns**: N/A

**Settings**: N/A

edx-webhook commented:
Thanks for the pull request, @jcdyer! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=16186&repo=edx%2Fedx-platform"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

bradenmacdonald commented on common/lib/xmodule/xmodule/assets/vertical/public/js/vertical_student_view.js:
```diff
@@ -13,4 +13,25 @@ window.VerticalStudentView = function(runtime, element) {
             apiUrl: $bookmarkButtonElement.data('bookmarksApiUrl')
         });
     });
+    console.log("jcd VerticalStudentView loaded");
```
Don't forget to remove these `console.log`s :)

bradenmacdonald commented on common/lib/xmodule/xmodule/assets/vertical/public/js/vertical_student_view.js:
```diff
@@ -13,4 +13,25 @@ window.VerticalStudentView = function(runtime, element) {
+            const data = {
+                'event_type': 'completion',
+                'block_key': block_key,
+		'completion': 1.0
```
Nit: inconsistent indentation here. Also, you can use trailing commas on the last element in JS arrays/objects, just like in python.

bradenmacdonald commented on common/lib/xmodule/xmodule/vertical_block.py:
```diff
@@ -1,27 +1,47 @@
+
+    """
+    return (
+        getattr(block, 'completion_method', 'completable') == 'completable'
```
Could we put `completable` into a constant, e.g. `from lms.djangoapps.completion import CompletionMethod`, then `CompletionMethod.completable` ?

bradenmacdonald commented on common/lib/xmodule/xmodule/vertical_block.py:
```diff
@@ -1,27 +1,47 @@
 
+def is_default_completable(block):
+    """
+    Returns True if the block can by completed by the default method.
```
"the default method" is pretty vague - perhaps we should say "the default method (marking it as completed once it has been viewed)" or something ?

bradenmacdonald commented on common/lib/xmodule/xmodule/vertical_block.py:
```diff
@@ -37,6 +57,27 @@ class VerticalBlock(SequenceFields, XModuleFields, StudioEditableBlock, XmlParse
 
+    def get_completables_for_user(self, username):
+        """
+        Return a set of descendent blocks that are completable.
```
This isn't an accurate description anymore - this is the set of descendent blocks that this vertical should mark as complete once the user views them. (i.e. it excludes completable blocks that are complete, and it excludes completable blocks that get marked complete via some other mechanism).

bradenmacdonald commented on common/lib/xmodule/xmodule/vertical_block.py:
```diff
@@ -66,7 +107,7 @@ def student_view(self, context):
 
             contents.append({
-                'id': child.location.to_deprecated_string(),
+                'id': unicode(child.location),
```
Nit: Or `six.text_type` if you're changing this anyways?

bradenmacdonald commented on common/lib/xmodule/xmodule/vertical_block.py:
```diff
@@ -76,7 +117,8 @@ def student_view(self, context):
             'bookmarked': child_context['bookmarked'],
-            'bookmark_id': u"{},{}".format(child_context['username'], unicode(self.location))
+            'bookmark_id': u"{},{}".format(child_context['username'], unicode(self.location)),
+            'completables': self.get_completables_for_user(child_context['username']),
```
How about calling this `watch_completables` or something to distinguish it from a list of children which are completable? See past comment about this list not being simply a list of completable descendants.

bradenmacdonald commented on common/lib/xmodule/xmodule/vertical_block.py:
```diff
@@ -177,3 +219,32 @@ def index_dictionary(self):
+            block_key = UsageKey.from_string(data.pop('block_key'))
+            block = self.find_descendent(block_key)
+            if block is None:
+                raise JsonHandlerError(400, "Invalid block: %s" % block_key)
```
Should we also check that this block is in the set returned by `get_completables_for_user` ? Or do we not care if spurious completion events or premature ones are emitted (e.g. user calling this directly to mark a gradable item as completed, instead of completing the graded activity).

jcdyer replied:
I think it doesn't matter if it's in the set of get_completables_for_user.  The main purpose of the specific user check is to reduce network traffic, and at this point, the request is already in the VPC (and will not even hit the DB if it doesn't change the value).

jcdyer replied:
Hmm.  But it might be worth the looser check of whether it uses the default completion method.  I'll think about that.

bradenmacdonald commented on common/lib/xmodule/xmodule/assets/vertical/public/js/vertical_student_view.js:
```diff
@@ -13,4 +13,25 @@ window.VerticalStudentView = function(runtime, element) {
+    console.log("jcd VerticalStudentView loaded");
+    $(element).find('.vert').each(function(idx, completable) {
+        console.log(completable.dataset);
+        if (completable.dataset.completable != "True") {
```
Nit: I'm not a fan of this string "True" comparison. How about putting the `data-completable` attribute (with no value) on the element if and only if it should be marked as completed on view (or perhaps call it `data-complete-on-view`?), and then change this check here to use `hasAttribute`?

i.e. https://stackoverflow.com/questions/16864999/can-i-use-html5-data-attributes-as-boolean-attributes

jcdyer replied:
Sounds good.  I don't love "stringly" typed data either.  I'll do it that way.

I suppose another option would be to make the values "JSON like" (so `data-bool='true' data-null='null' data-num='1' data-string='"string"' data-object='{"tweedle": "dee"}'`), but you'd need to make that a site-wide policy to be worth doing anyway.  (just musing out loud).

jcdyer commented:
Wow.  I got my git branches in a muddled state.  I think everything's back in shape now.

bradenmacdonald commented on lms/djangoapps/completion/tests/test_handlers.py:
```diff
@@ -0,0 +1,88 @@
+            score_db_table='submissions',
+        )
+        completion = BlockCompletion.objects.get(user=user, course_key=course_key, block_key=block_key)
+        self.assertEqual(completion.completion, 1.0)
```
This test doesn't seem to be doing what it says (`test_handler_resets_completion_on_score_delete`) and it seems redundant given the `ddt` variants of the `test_handler_submits_completion ` test above. Is this just a remnant from before a ddt refactor?

jcdyer replied:
Yeah.  The DDT was a late addition, and I think I did it half way through copy-pasting this test (before revising it to match the name.

bradenmacdonald commented:
If you want to spin the `has_custom_completion` nuances out into a separate story, in the interest of merging and cleaning up, however, I'd be fine with that.

jcdyer commented:
@bradenmacdonald You're absolutely right about checking for `has_custom_completion`.  I've created a new ticket ([OC-3247](https://tasks.opencraft.com/browse/OC-3247)), and added it to the epic.

I also removed the duplicate test.

bradenmacdonald commented:
@robrap 

> In the future, would it be possible to base the branch and PR off the other WIP branch before it merges, so the PR just has the changes meant to be reviewed?

That would be really nice, yep - especially since GitHub allows changing the target branch before merging a PR, so it can change from the previous PR to master before merging. However, I don't think @jcdyer has push access to this repo, and it would be too inefficient for you or I to keep pushing his changes from the `open-craft` fork to here. If he could be given push access for at least the duration of this project, it would make the PRs much more manageable :)

robrap commented:
@jcdyer @bradenmacdonald: Regarding my [earlier comment](https://github.com/edx/edx-platform/pull/16186#pullrequestreview-69903070), I submitted a ticket to get @jcdyer push access.  I'll let you know when I hear back.  Point 2 is still open and waiting for a response.
Thanks.

jcdyer commented:
Good point, @robrap.  Nothing is going to stop the block completions from being written.  I was planning on allowing the BlockCompletions to be written, and then using COMPLETION_START_DATE to prevent using early BlockCompletions from affecting AggregateCompletions (once those are implemented, but it probably makes sense to have basic completion rendering in place, and relatively complete, so the aggregate code doesn't have to be quite as defensive.  I'll introduce a waffle flag in #16112.

jcdyer commented:
Thanks for looking into getting me push access.  Additionally, I'm hoping once these first two commits land (model and event handler), I won't need to stack PRs as much.

robrap commented:
@jcdyer: I still haven't thought this through, but I imagine it would be simpler if for any user, the timestamp of any completion data is the point at which that user has full completion data (e.g. default visibility completion, etc.).  I guess we could truncate the table if and when we really wanted to achieve the same, but if the waffle flag were a part of this PR, we wouldn't get any data until this were turned on.

Also, maybe this is a non-issue, but do we have any known scored blocks that will not track completion once the custom completion and exemption code is done?  In other words, will we be writing bad data in certain cases with this PR?

Lastly, it sounds like push access will be coming soon, but not certain until you have it. :)

jcdyer commented:
jenkins run lettuce

robrap commented:
@jcdyer: You now have push access.  A couple of thoughts:
1. I will try not to hold up any PRs, but please try to avoid merging until both @bradenmacdonald and I have approved.
2. I see that I haven't approved this ticket yet, but also see it includes issues that were written/updated on another ticket (I think), and I'm confused about what is needed on this one.  Are you able to clean up the PR, basing it on another branch, and/or reminding me to review this one after the dependent changes are approved or merged and updated or removed from this PR?
Thanks.

jcdyer commented:
@robrap 

> 1. I will try not to hold up any PRs, but please try to avoid merging until both @bradenmacdonald and I have approved. 

Of course.

> 2. I see that I haven't approved this ticket yet, but also see it includes issues that were written/updated on another ticket (I think), and I'm confused about what is needed on this one. Are you able to clean up the PR, basing it on another branch, and/or reminding me to review this one after the dependent changes are approved or merged and updated or removed from this PR?
>
> Thanks.

This one depends on the changes in cliff/handle-block-completion (#16112).  It was getting time consuming to keep rebasing this and #16234 whenever I added a commit to that PR, so it may be a little out of data ATM.  Once that gets merged, I'll update this to match it, ensure all outstanding review issues are addressed, and ask for another review.

robrap commented:
Great.  I'll wait for your prompt on this one.  Thanks.

jcdyer commented:
@robrap @bradenmacdonald This has been rebased against the master branch.  Can I get a fresh review on it?  I had to expand the waffle flag handling a bit to prevent breaking grading, and I accidentally did it during the rebase cleanup, so it's all still one commit.

robrap commented:
@jcdyer: I can't find your earlier comment from one of these PR's with a "Xsoon" joke, but I appreciated it.

jcdyer commented:
@robrap Glad to hear it.  I don't have any particular concerns on the latest implementation.  I'll get this squashed and merged.
